### PR TITLE
add usb-raw iterface

### PR DESCRIPTION
**Pull request by teknoraver**

### CODE REVIEW

File: interfaces/builtin/rawusb.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+package builtin
+
+import (
+	"github.com/snapcore/snapd/interfaces"
+)
+
+const rawusbConnectedPlugAppArmor = `
+# Until we have proper device assignment, allow access to all usb devices
+/dev/bus/usb/[0-9]{3}/[0-9]{3} rw,
```
Comment (jdstrand) [inline interfaces/builtin/rawusb.go]: This rule does not compile. Please adjust. I suspect you are trying to use a regex here but apparmor syntax is a bit different. I suspect this will do what you want:
```
/dev/bus/usb/[0-9][0-9][0-9]/[0-9][0-9][0-9] rw,
```

File: interfaces/builtin/rawusb.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+# Allow detection of usb devices. Leaks plugged in USB device info
+/sys/bus/usb/devices/ r,
+/sys/devices/pci*/*/usb*/**/idVendor r,
+/sys/devices/pci*/*/usb*/**/idProduct r
```
Comment (jdstrand) [inline interfaces/builtin/rawusb.go]: This last rule will cause problems because of the missing comma, but otherwise these rules are fine, can we adjust this to be though:
```
/sys/devices/pci*/**/usb[0-9]*/**/idVendor r,
￼/sys/devices/pci*/**/usb[0-9]*/**/idProduct r,
```

File: interfaces/builtin/rawusb.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+`
+
+// NewRawUsbInterface returns a new "rawusb" interface.
+func NewRawUsbInterface() interfaces.Interface {
+	return &commonInterface{
+		name: "raw-usb",
```
Comment (jdstrand) [inline interfaces/builtin/rawusb.go]: We have a 'hiddraw' interface (I realize that is a name from the kernel), so 'raw-usb' sounds a bit inconsistent to me. What about 'usbraw'? @niemeyer, can you also weigh in on the naming of this transitional interface?

File: interfaces/builtin/rawusb.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+package builtin
+
+import (
+	"github.com/snapcore/snapd/interfaces"
+)
+
+const rawusbConnectedPlugAppArmor = `
+# Until we have proper device assignment, allow access to all usb devices
```
Comment (jdstrand) [inline interfaces/builtin/rawusb.go]: Can you reword this to be: `Transitional interface which allows access to all usb devices`

File: snap/implicit.go [hunk 5]
```diff
@@ -57,6 +57,7 @@ var implicitSlots = []string{
 	"tpm",
 	"kernel-module-control",
 	"camera",
+	"raw-usb",
```
Comment (jdstrand) [inline snap/implicit.go]: I realize this section is also not alphabetized, can you alphabetize it?

File: interfaces/builtin/rawusb.go (hunk 2 above)
Comment (teknoraver) [inline interfaces/builtin/rawusb.go]: isn't `pci**/` better than `pci*/**/`?

File: interfaces/builtin/rawusb.go (hunk 3 above)
Comment (teknoraver) [inline interfaces/builtin/rawusb.go]: rawusb is just a full usb access, to handle all the protocol in userspace (eg. with libusb)
that's actually what some programs does, like sane

File: interfaces/builtin/all_test.go [hunk 6]
```diff
@@ -48,6 +48,7 @@ func (s *AllSuite) TestInterfaces(c *C) {
 	c.Check(all, Contains, &builtin.PulseAudioInterface{})
 	c.Check(all, Contains, &builtin.UDisks2Interface{})
 	c.Check(all, Contains, &builtin.FwupdInterface{})
+	c.Check(all, Contains, &builtin.RawusbInterface{})
```
Comment (jdstrand) [inline interfaces/builtin/all_test.go]: This commit is fine-- it is unrelated, but easy to follow.

File: interfaces/builtin/rawusb.go (hunk 2 above)
Comment (jdstrand) [inline interfaces/builtin/rawusb.go]: Yes, that's fine.

**Comment by teknoraver** [comment]
@jdstrand there is a test which I can't figure out:
```
----------------------------------------------------------------------
FAIL: basedeclaration_test.go:125: baseDeclSuite.TestAutoConnection

basedeclaration_test.go:169:
    c.Check(err, NotNil, comm)
... value = nil
... usb-raw

OOPS: 361 passed, 1 FAILED
--- FAIL: Test (0.04s)
FAIL
exit status 1
FAIL    github.com/snapcore/snapd/interfaces/builtin    0.040s
```
how to fix it? it seems related to autoconnect

**Comment by teknoraver** [comment]
@jdstrand there is a test failure which probably is unrelated to this commit

File: docs/interfaces.md [hunk 7]
```diff
@@ -518,3 +518,9 @@ to storage on the device
 Can query UPower for power devices, history and statistics.
 
 * Auto-Connect: yes
+
+### usb-raw
+
+Can send and reveice data directly to any USB device attached.
```
Comment (jdstrand) [inline docs/interfaces.md]: Typo and grammar issues. Please use: `Can directly send and receive data to an connected USB device`

File: interfaces/builtin/all.go [hunk 8]
```diff
@@ -40,45 +41,45 @@ var allInterfaces = []interfaces.Interface{
...
-	NewOpenglInterface(),
-	NewCupsControlInterface(),
-	NewOpticalDriveInterface(),
-	NewCameraInterface(),
-	NewBluetoothControlInterface(),
-	NewKernelModuleControlInterface(),
-	NewFuseSupportInterface(),
-	NewShutdownInterface(),
-	NewTimeControlInterface(),
-	NewLibvirtInterface(),
```
Comment (jdstrand) [inline interfaces/builtin/all.go]: Thank you for alphabetizing these.

**Comment by morphis** [comment]
@jdstrand Can you have another look at this?

**Comment by jdstrand** [comment]
Sorry, I meant to approve the changes yesterday.

**Comment by jdstrand** [comment]
I think @niemeyer should give a +1 on the name, but otherwise +1 from me.

File: interfaces/builtin/usb_raw.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+# Description: Allow raw access to all connected USB devices.
+# Reserved because this gives privileged access to the system.
+# Usage: reserved
+/dev/bus/usb/[0-9][0-9][0-9]/[0-9][0-9][0-9] rw,
+
+# Allow detection of usb devices. Leaks plugged in USB device info
+/sys/bus/usb/devices/ r,
+/sys/devices/pci**/usb[0-9]**/idVendor r,
+￼/sys/devices/pci**/usb[0-9]**/idProduct r,
+`
```
Comment (swem) [inline interfaces/builtin/usb_raw.go]: @teknoraver Could you help to add below apparmor rules to this usb-raw interfaces. This would be help for Intel Realsense applications. See my pull request for realsense interface https://github.com/snapcore/snapd/pull/2236#pullrequestreview-6607126 . If these rules being intergrated into usb-raw interfaces. Then I can leverage usb-raw interface directly and will not need to add another interface for realsense applications.

/sys/devices/pci**/usb*/**/busnum r,
/sys/devices/pci**/usb*/devnum r,
/sys/devices/pci**/usb*/**/devnum r,
/sys/devices/pci**/usb*/**/modalias r,
/sys/devices/pci**/usb*/**/bInterfaceNumber r,

Comment (morphis) [inline interfaces/builtin/usb_raw.go]: @jdstrand Can you comment on this from the security perspective?

Comment (teknoraver) [inline interfaces/builtin/usb_raw.go]: why do you access data via /sys/bus/pci**/usb** instead of /sys/bus/usb**/?
it seems that the information is in both tree
```
$ ll /sys/bus/usb/devices/*/*/modalias
-r--r--r-- 1 root root 4,0K nov  2 10:59 /sys/bus/usb/devices/1-11/1-11:1.0/modalias
-r--r--r-- 1 root root 4,0K nov  2 10:59 /sys/bus/usb/devices/1-4/1-4:1.0/modalias
-r--r--r-- 1 root root 4,0K nov  2 10:59 /sys/bus/usb/devices/1-8/1-8:1.0/modalias
-r--r--r-- 1 root root 4,0K nov  2 10:59 /sys/bus/usb/devices/1-8/1-8:1.1/modalias
-r--r--r-- 1 root root 4,0K nov  2 10:59 /sys/bus/usb/devices/1-8/1-8:1.2/modalias
-r--r--r-- 1 root root 4,0K nov  2 10:59 /sys/bus/usb/devices/1-8/1-8:1.3/modalias
-r--r--r-- 1 root root 4,0K nov  2 10:59 /sys/bus/usb/devices/2-2/2-2:1.0/modalias
-r--r--r-- 1 root root 4,0K nov  2 10:59 /sys/bus/usb/devices/2-4/2-4:1.0/modalias
-r--r--r-- 1 root root 4,0K nov  2 10:59 /sys/bus/usb/devices/3-1:1.0/0003:046D:C404.0001/modalias
-r--r--r-- 1 root root 4,0K nov  2 10:59 /sys/bus/usb/devices/3-1/3-1:1.0/modalias
-r--r--r-- 1 root root 4,0K nov  2 10:58 /sys/bus/usb/devices/usb1/1-0:1.0/modalias
-r--r--r-- 1 root root 4,0K nov  2 10:58 /sys/bus/usb/devices/usb2/2-0:1.0/modalias
-r--r--r-- 1 root root 4,0K nov  2 10:58 /sys/bus/usb/devices/usb3/3-0:1.0/modalias
-r--r--r-- 1 root root 4,0K nov  2 10:58 /sys/bus/usb/devices/usb4/4-0:1.0/modalias
$ ll /sys/bus/pci/devices/*/usb*/*/modalias
-r--r--r-- 1 root root 4,0K nov  2 10:58 /sys/bus/pci/devices/0000:00:14.0/usb1/1-0:1.0/modalias
-r--r--r-- 1 root root 4,0K nov  2 10:58 /sys/bus/pci/devices/0000:00:14.0/usb2/2-0:1.0/modalias
-r--r--r-- 1 root root 4,0K nov  2 10:58 /sys/bus/pci/devices/0000:04:00.0/usb3/3-0:1.0/modalias
-r--r--r-- 1 root root 4,0K nov  2 10:58 /sys/bus/pci/devices/0000:04:00.0/usb4/4-0:1.0/modalias
```

Comment (teknoraver) [inline interfaces/builtin/usb_raw.go]: I see it, they are all links to `./../../devices/pci0000:00`

Comment (teknoraver) [inline interfaces/builtin/usb_raw.go]: @swem is this ok?
```
/sys/devices/pci**/usb[0-9]**/busnum r,
￼/sys/devices/pci**/usb[0-9]**/devnum r,
/sys/devices/pci**/usb[0-9]**/modalias r,
/sys/devices/pci**/usb[0-9]**/idVendor r,
￼/sys/devices/pci**/usb[0-9]**/idProduct r,
￼/sys/devices/pci**/usb[0-9]**/descriptors r,
```

File: docs/interfaces.md [hunk 10]
```diff
@@ -518,3 +518,9 @@ to storage on the device
 Can query UPower for power devices, history and statistics.
 
 * Auto-Connect: yes
+
+### usb-raw
+
+Can directly send and receive data to a connected USB device.
```
Comment (jdstrand) [inline docs/interfaces.md]: This doesn't limit to a particular device, so can we reword to be `Can directly send and receive data to connected USB devices.`

Sorry I didn't think of this before.

File: interfaces/builtin/usb_raw.go (hunk 9 above)
Comment (jdstrand) [inline interfaces/builtin/usb_raw.go]: @morphis, this is a powerful interface but the requested additional access are all related to usb devices and read-only, so yes, these are ok. That said, seeing the udev rule suggests to me that we should change `/run/udev/data/c189:[0-9] r, # USB serial converters` to:
```
/run/udev/data/c16[67]:[0-9] r, # ACM USB modems
/run/udev/data/b180:[0-9]    r, # various USB block devices
/run/udev/data/c180:[0-9]    r, # various USB character devices
/run/udev/data/c18[89]:[0-9] r, # USB serial converters
```

Comment (teknoraver) [inline interfaces/builtin/usb_raw.go]: it seems that many other files are needed, this are the denials from usb-devices:
```
matteo@matteo-desktop:~/src/canonical/snapd$ usb-utils.usb-devices &>/dev/null
matteo@matteo-desktop:~/src/canonical/snapd$ dmesg |awk -F'"' '/apparmor="DENIED"/{print$8}'
/sys/devices/pci0000:00/0000:00:14.0/usb1/maxchild
/sys/devices/pci0000:00/0000:00:14.0/usb1/version
/sys/devices/pci0000:00/0000:00:14.0/usb1/bDeviceClass
/sys/devices/pci0000:00/0000:00:14.0/usb1/bDeviceSubClass
/sys/devices/pci0000:00/0000:00:14.0/usb1/bDeviceProtocol
/sys/devices/pci0000:00/0000:00:14.0/usb1/bMaxPacketSize0
/sys/devices/pci0000:00/0000:00:14.0/usb1/bNumConfigurations
/sys/devices/pci0000:00/0000:00:14.0/usb1/bcdDevice
/sys/devices/pci0000:00/0000:00:14.0/usb1/bcdDevice
/sys/devices/pci0000:00/0000:00:14.0/usb1/manufacturer
/sys/devices/pci0000:00/0000:00:14.0/usb1/speed
```
what about just allowing `/sys/devices/pci**/usb[0-9]** r,`?

File: interfaces/builtin/usb_raw.go [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+# Allow detection of usb devices. Leaks plugged in USB device info
+/sys/bus/usb/devices/ r,
+/sys/devices/pci**/usb[0-9]**/busnum r,
+/sys/devices/pci**/usb[0-9]**/devnum r,
+/sys/devices/pci**/usb[0-9]**/modalias r,
+/sys/devices/pci**/usb[0-9]**/idVendor r,
+/sys/devices/pci**/usb[0-9]**/idProduct r,
+/sys/devices/pci**/usb[0-9]**/descriptors r,
+`
```
Comment (jdstrand) [inline interfaces/builtin/usb_raw.go]: @teknoraver asked this:
"what about just allowing `/sys/devices/pci**/usb[0-9]** r,`?"

Considering the nature of this interface, please feel free to use `/sys/devices/pci**/usb[0-9]** r,` instead of all these /sys/devices/pci** rules

File: interfaces/builtin/usb_raw.go (hunk 9 above)
Comment (teknoraver) [inline interfaces/builtin/usb_raw.go]: also, there are devices with three digits, I'd use `/run/udev/data/c18[089]:* r,`

**Comment by teknoraver** [comment]
Rebased and fixed the failing legacy autoconnect test

**Comment by niemeyer** [comment]
The name seemed fine originally. hidraw comes from the kernel indeed, and perhaps was not ideally named, but it's a single term such as pulseaudio. raw is not part of usb, though, and our naming convention has adjective before noun elsewhere.

**Comment by niemeyer** [comment]
@teknoraver Sorry for the trouble, but can you please submit a branch inverting the name back to what you had before?

**Comment by teknoraver** [comment]
@niemeyer ok, so "rawusb"?

**Comment by niemeyer** [comment]
@teknoraver "raw-usb" please..  hidraw is a confusing special case because it's a noun [in practice](https://www.kernel.org/doc/Documentation/hid/hidraw.txt). Much like "pulseaudio", which isn't about "pulsing audio". "usb" is a term on its own, though, and "raw" is an adjective here, rather than a part of that name. Thus, "raw-usb".

**Comment by teknoraver** [comment]
created #2297
